### ISSUE TITLE
Create Complete Uchaguzi Dataset for COMRADES use

### ISSUE THREAD

**Issue by Shadrock** [ushahidi/platform#3073]
## Overview
The output of this Epic (perhaps just an issue?) will be a complete dataset of Uchaguzi data in .csv format. While there have been several Uchaguzi datasets downloaded for COMRADES all of them have been difficult to use (requiring concatenation; confusing headers; etc.) and we need a _complete_ and unified dataset for evaluations. 

## Who is it for? Who are the users?
This is for use by COMRADES evaluation analysis including Shadrock. It could also be used as a historical dataset by Hazel or others in Ushahidi.

## What privacy impacts does it have?
- This request does _not_ change the personal data we collect.
- This dataset may involve data that was previously private by being either unpublished (archived) or marked as private in the deployment. However, this dataset is for COMRADES use only. Any posts that were to be published (e.g. as an example in a report) are required to have PII scrubbed from them by the rules governing the COMRADES grant.

## Requirements
A "complete" Uchaguzi dataset means that data for both deployments will be captured and include those posts that were "archived" or otherwise marked as unusable. This is important because, for the evaluations, we need to look at rations of usable to non-usable information and perform other sorts of machine learning. The data from both deployments can be combined into one dataset if there is a method to differentiate between records for Uchaguzi 1 and 2Chaguzi (e.g. if the posts can be filtered by a date and the that date does not overlap between both deployments), but I would prefer to see them as two separate .csv files.

Furthermore, for better analytics, the dataset should follow certain guidelines, possibly including:
- The dataset should not contain loads of empty or extraneous empty cells. 
- The dataset should have clearly named headers and/or a detailed explanation of header codes. 
- Each record in the dataset should include any relevant form fields for the associated post and *include* what survey the post was attached to; any tags attached to the post; and whether or not the post was "escalated" (which I believe has been a variable in all subsequent data sets I have seen). 

### Who will this be deployed to?
e.g. will it be feature flagged? will it go to just once deployment?

## Deployment

*Does this deploy new services?*
*Do the new services have health checks?*
*Can this be feature flagged and deployed disabled?*

**Comment by jrtricafort** [comment]
@willdoran Looping you in on this. Let's discuss here this week.

**Comment by jrtricafort** [comment]
1) We (Platform team) @willdoran will do a CSV export of the entire Uchaguzi data with the newly-corrected columns. We will do one export that includes all surveys, as well as separate exports for each survey type.

2) Share with @Shadrock and have conversation about what else he needs of that CSV.
Possible outcomes and actions:
2a) CSV now matches what Shadrock needs -> Woohoo
2b) CSV contains additional unneeded data -> Manually clean/trim the CSV
2c) CSV is missing data or data is mis-formatted -> Would require more CSV changes, and that would 
2d) Some further data manipulation is required of either COMRADES or us

**Comment by jrtricafort** [comment]
Next steps:

@willdoran to provide @Shadrock with current CSV format. 

Then @Shadrock will let us know what, if anything, needs to be changed in the format for this to be acceptable.

**Comment by willdoran** [comment]
@Hazellaura @Shadrock Here is the CSV to be reviewed
https://drive.google.com/open?id=1V3wjESA0hAW5GjobHE6FkeRlo_vzvSsC

**Comment by jrtricafort** [comment]
@willdoran uploaded the CSV of Uchaguzi data in the comment above^^

@Shadrock Can you let us know what changes, if any, need to be made to the format prior to providing this to COMRADES and closing out this issue?

@Hazellaura Can you let us know if you have feedback on the format? Feel free to comment on the document or here (but be careful not to share actual data on this issue). Depending on how this overlaps with Shadrock's feedback, we may incorporate those changes for COMRADES or take that work up as a separate ticket as part of this epic: https://github.com/ushahidi/platform/issues/3119

**Comment by Hazellaura** [comment]
@willdoran Please find my comments on the CSV file that you sent. let me know if you need me to clarify and of my points. Also, I have attached a CSV download from a specified survey (Gender Based Violence) from one of our DREAMS deployment. I made a comment in the CSV you forwarded and the DREAMS CSV will illustrate my point further @Shadrock @jrtricafort @rowasc , I hope this is useful. 
[Hazel's Copy of csv-export-2018-06-28-5b35271f6f57e-full.csv.xlsx](https://github.com/ushahidi/platform/files/2155209/Hazel.s.Copy.of.csv-export-2018-06-28-5b35271f6f57e-full.csv.xlsx)

[ACWICT GBV Survey 3.xlsx](https://github.com/ushahidi/platform/files/2155229/ACWICT.GBV.Survey.3.xlsx)

**Comment by willdoran** [comment]
@Hazellaura Sorry I should have said not to share data here as it's public, I've downloaded your docs though and I'll copy the comments here.

Thank you.

**Comment by Hazellaura** [comment]
I am so sorry Will.

**Comment by willdoran** [comment]
### Columns: 
#### Survey name
This is an important column. I think it was previously form name like Will mentioned.
#### Post ID
This is also an important column. It helps findings specific posts.
#### Sets
I usually never use the sets options. Not quite sure what sets mean. I would do away with it.
#### Description
The description field is really confusing to Users based on our DREAMS Partners feedback. It is not clear what description is but based on the CSV file export, the description is the content of the post or message. I wish we could just name this Post content or message content.
#### Title
The title as well is not clear. However, after the structuring process, the Title ends up being the message content summary or a portion of the message. I would also recommend we changed this and make it more clear. Perhaps. Message summary, or summary of post content. We can discuss this further.
#### Incident date
I don't understand the incident date. Is tit different from the content, updated and post dates? If not, I would remove this. If the incident date however is the date of a specific incident e.g. Violence like during elections, this would be useful. But I doubt this, assume the incident date is the same as the Post date. I would also consider removing this.
#### Location.lat
Again this is not useful for data analysis.
#### Nature of this incident
This was really useful for the Uchaguzi data. These field went into coding the posts that we received according to different themes. This could be a duplicate of the title where the content of the post is summarised according to a theme. But this was really useful. It is a manual process though and requires structuring and analysis to begin on platform. It was super useful.
#### Image of incident
I did not use this column at all. I would select images straight from platform where I needed an image. I would do away with this actually.
#### Video link
I would also do away with this.
#### Transition status
I did not use this much. This field would depend on what the person analysing data is trying to report. Also, this field would be useful if I wanted to follow up on what posts needed to be translated and why they have not yet been translated e.t.c. I would not need this column here. In fact,  if I needed to see translation status data, I could perhaps download only this 
#### Related Posts
I think this may be useful but was not available during Uchaguzi, so I never used it. I would keep it though.
#### Title
Title 2: Here the CSV Starts again. This is happening with DREAMS CSV's as well. This is something that has just started to happen. Previously, the field would repeat themselves depending on the different surveys ie. the data for individual surveys would have their own columns for each row title. This meant cleaning the data into one similar column (I hope this makes sense). 
I though that downloading one survey at a time would help, however, it seems like doing this complicates the CSV further in that it includes items from different surveys into the specified survey and you end up with blank columns again because these added items are not specific to the  selected survey.

**Comment by willdoran** [comment]
@Hazellaura Don't worry at all, we should have been super clear about that it was our fault.

**Comment by willdoran** [comment]
### Comments
This was hugely useful for a number of reasons. I have dropped out some of the comments as they were either - this is very useful or this was not useful at all.

1. Most importantly there is a miscommunication between platform and the PEOs about the structure of the CSV. It's not totally clear to me why we have failed to explain this properly but we need to address it quickly.

If you export multiple Surveys in the same CSV file you will see Post Id, Survey and then all the columns for the first Survey in order, then once those are finished you will see all the columns for the next Survey and so on. 

The fields are never being repeated it is just that between Surveys there are fields that will always or may have the same name so for example. I have 3 surveys, my CSV structure will be

Post ID, Survey...Sets, Title, Desription, ...<fields for survey 1>, Title, Description...Fields Survey 2....Title, Description,...<fields survey 3>

The fields Title and Description appear exactly 3 times, each time they are for a different Survey. The reason we included the column named Survey that holds the Survey name is to allow grouping of the columns by Survey to make it easier to see which columns belong to which Surveys.

Sometimes - for example with Uchaguzi - many Surveys have field names that are exactly the same, however, to the system these are completely different fields even though they have the same names.

We are very limited in what we can do about this. The options we have are:
1. Always export each Survey in its own file.
2. Provide an option to allow the user to choose if they want 1 file or 1 file per Survey
3. We can change the column names from Title to something like Survey1.Title. That was, however, the original structure we used but it was removed because it made the column names difficult to deal with.

A big part of this issue comes from the fact that by putting multiple Surveys in a single file we are trying to relate structures that were never designed to be directly compared.

One of the major advantages of placing the information in 1 file is that Survey Posts do have overlapping fields that may be used to track information across all the Surveys, otherwise it would be very hard to do any kind of statistical analysis between Surveys.

One suggestion in the past was to create global fields that are shared across Surveys and can be essentially reused. I think this is probably one of the best approaches but we will have to make the names of Survey fields globally unique which will be difficult to do for existing instances of platform.

I would like to check that we all have a shared understanding of the content of the CSV files so I think it might be useful for us to discuss one in a call where we can explain live the fields and why they appear and @Hazellaura could tell us how the CSV will be used for actual analysis.

2. We need to put the Type of each field on the Survey export page not just it's name. If I'm exporting data there is no guarantee I was the person who made the Survey so ALL the information necessary to understand what I am actually exporting should be displayed on the export page otherwise I will always just export everything and delete what I don't need making the field selector irrelevant.

If I am doing stats there is one set of field types I care about. If I'm doing qualitative there is another and if I'm prepping data for machine learning there is a completely different set.

3. If I select a complex arrangements of fields to export I probably want to be able to reuse that later, we already save this information so we should consider adding a "Rerun export button" on the history page which will copy the previous export and run it again with the same filters/fields

4. The naming scheme for columns needs to be better done but we need to communicate that the data exported is in the rawest form imaginable and that it will be ready for Data Cleaning BUT will always require some form of work to get it into the state that a particular group needs and I believe we should not artificially combine columns, we can't handle all possible usecases for this. However, we definitely need to make sure the data is in a clear format. 

I would like to ask @Hazellaura some more specific questions around the formatting of special fields like text and date.

**Comment by Shadrock** [comment]
### Executive Summary (detailed feedback below): 
@jrtricafort @willdoran Here's my feedback. There are two major showstoppers before a CSV structed like this could be used by COMRADES. 

- Each row contains multiple records (or appears to). I must have each row showing a single record (or ‘post’). I discuss this in more detail below.
- Several of the "status" variables (e.g. “Translation Status”) re-appear as the first value in the record (e.g. “Translation Status, Translated”). The header should *not* re-appear in the cells below it. Cells should contain unique values only. If every value in a dataset starts with the column header than the extra (and extremely annoying) step of removing that header name must happen before the data can be used. In some cases the heading names are confusing, but this can be dealt with later.

Documentation about what, exactly, the headers mean is also an issue, but not a showstopper. One final question: judging by *most* of the values in the "Created" column I would guess that these data were for the Uchaguzi 1. I will need data from both deployments and need a way to distinguish between them. If I can do that by one of the (many) date values, then one dataset is fine... but it may be better to break them up into two datasets, one for each period of voting. Finally, why am I seeing dates like 2/7/2018 in this data set? Where the deployments receiving posts months after the fact? 

### Detailed Feedback: 
Note that I did *not* review Hazel's comments before writing these notes, so anything that we both notice is likely a real problem. As I went through this file I realized that I was evaluating this CSV on a few different levels. The first was to look at it as a unique document that I would try to use as an analyst. This feedback is fine-grained, but very specific to the way in which the Uchaguzi deployment was set up and specific to this issue. The other way I reviewed this file was as a general format for a CSV exported from a deployment, which indicates issues larger than COMRADES work. Both of the showstoppers listed above fall, I believe, into the latter category. 

The immediate problem with this file is that it will put-off a lot of analysts because it duplicates headers. Example: Column J = Description, Column AE = Description, AW = Description, etc.. I have no idea why this is and it’s incredibly confusing. Each row should be a singular record. Rows should not contain multiple records. This is required for COMRADES work and should be the norm for all exports IMHO. If there’s a logic for having each row contain multiple records, I don’t understand it. For the rest of the review I will examine *one set* of unique column headings (e.g. A - AC if viewed in a spreadsheet). 

Another issue is documentation. I realize that column headers that are the product of configurations by users are not something we can manage. However, for any “stock” variables that will show up in an export we need to provide clear documentation on what they are. See [my example here for RNI financial data](https://github.com/Shadrock/RNI-Activity-Expenses). I have listed all the key codes so a user can see what they mean and how they were collected. 

### Detailed Feedback by Column Header: 
This is a mix of observations about the general format of the csv, but also notes concerning variables specific to Uchaguzi. As such, they are mostly related to naming and probably don't represent show stoppers (unless already mentioned above). They may or may not be useful for wider discussions around .csv exports. 

**Post ID -** this is fine but there needs to be documentation about what it is. I’m assuming it’s a number assigned in order as posts come in. Thus, if I see a single record from a deployment labeled “61,009” then I know they at least received >61,000 posts. Is this correct? 

**Survey & Post status -** these both look good (although what a “draft post” is will need to be explained - is it incoming, published then in the process of being edited?). What I like with these two fields is that I can quickly use them together to get a feeling for which posts where considered useful. This is incredibly important for COMRADES, which requires bad data for training purposes just as much as good data. So this is an example of a really helpful format which gives me a unique variable related to a post that, when combined with other variables, tells me something important about the post. 

**Created (UTC) -** Fine but is this the date it hit the deployment, the date it was posted to Twitter (the ‘original’ created date) or what? Again, documentation necessary.

**Updated (UTC) -** No idea what this means. If it’s the last update, is that different from the Post date? 

**Post Date (UTC) -** Confusing. If we are dealing with “posts” then I would read “post date” as the same as “created.” I would assume this is the date the post came in. If we are talking about the date the post was *posted* (e.g. assigned a status) then this name should change. Even more so since I see that even archived posts get a “post date”… and I wouldn’t call archiving them, “posting them.”

**Contact ID -** No idea what this is. 

**Contact -** I guess this is the social media handle or the account name of a… user? The author of the post? Needs a lot of clarification here. Also, how do you differentiate (can I?) Between Twitter “contacts”, e-mail “contacts”, SMS phone number “contacts.”?

**Sets -** No idea what this is. 

**Description -** I understand what this is… but the name is confusing. I don’t think a lot of people would immediately understand what this is. This is the body of the post: it’s reported or scrapped content. Calling it a “description” leads me to believe that the actual content is elsewhere and this is just describing it. 

**Title -** As with “description” I know what this is and it makes sense since I’ve seen it in Ushahidi surveys… but now that I’m thinking about it… why do we even have titles? The posts are always short and should say it all. The assigned status and tags would be the more efficient way to “see what a post was about.” *We should consider actually doing away with titles IMHO.* 

**Incident Date -** this looks like a vestige of V2 that doesn’t make much sense here. In the old days, you’d search the body of the content for the date and time something happened. If we still do that, fine, but this needs to clearly indicate that. Seeing it here made me wonder how it related to the other three dates previously listed. 

**Location.lat -** WOOOHOO! It’s in it’s own column and it makes sense! 

**Location.lon -** WOOOHOO! It’s in it’s own column and it makes sense! 

**Refine Location Name -** What on Earth is this? Is it a manually assigned place name? If so, who assigned it? Confusing.

**Nature of this incident -** No idea what this is. Why is this different from survey status? 

**Image of incident -** Fine. 

**Video Link -** Fine.

**Translation Status -** Fine. It’s interesting to see if a post went through this particular workflow. However, the values are confusing. Each record that has a value in this column starts with “Translation.” That’s not needed if this is the “translation status” variable. Look at 35040 and 35042 for examples of how what’s being communicated is not quite clear. 

**Description Translation -** This is also confusing and, moreover, I think the values that should be assigned to it are getting placed into the translation status column. Again, see 35040 for an example: multiple values in “translation status” but non
 
**Geolocation Status -** While it’s good to see the process of what went into publishing a post, this is confusing. The status for 35040 is “Needs geolocation,Geolocation” but it clearly has lat/lon and, even, a “refined location name.” If this variable can be used to show that a location was extrapolated by somebody on the team running the deployment, versus having come from some automated method (e.g. GPS via a post made on mobile app) then this useful to have but could be communicated better. 

**Verification Status -** Good.

**Action Taken to Verify -** Good.

**Other posts related to this incident -** Good, and useful. But is it ever used? 

**Notes -** fine.

**Escalation Status -** fine 

**Action taken -** fine

**Escalation Status -** fine

**Media Stream Source -** Good. But, again, why does each value start with “media stream source”? This just means an analyst is going to have to strip that out before using it or before it displays properly in HDX. Each cell should only contain a value!

**Comment by Shadrock** [comment]
After reading @Hazellaura's comments (which are great...) I want to be clear that, for this Github issue, I do not want to redact columns just because they were not useful in the context of analyzing things for DREAMS. COMRADES is looking at the whole picture of data during the elections, which includes all the different fields created in the deployment, whether they were later found useful for not.

**Comment by willdoran** [comment]
@Shadrock Thank you for the level of detail and the time you took to cover this. Combining both your and Hazel's comments I am concerned that the actual system itself is failing to communicate what it is able to do and what is actually it's output.

I'll review the comments now.

**Comment by jrtricafort** [comment]
Thanks @Shadrock and @Hazellaura . 
@willdoran and I just connected on this. Next steps:

**Near Term **
Confirm scope of what we're doing to close out the COMRADES deliverable.
Shadrock, Will will send you an invite for this Friday for us to discuss and confirm.

**Medium Term**
Translate Hazel and Shadrock's feedback into both of the following:
(1) CSV feature changes
(2) Data model changes 
(2) Best practices for survey design processes (for Support and Implementation PEOs)

**Comment by Hazellaura** [comment]
I agree with Shadrocks comments concerning the focus not being on solely on
DREAMS for this Github issue. Because DREAMS and Uchaguzi are the only
reference points on the use of CSV's that I have, I agree my comments tend
to lean on these specific deployments. Let me rephrase, when I say
something may not be useful, I mean use in the context of the user. When I
am analysing data to report findings and communicate findings, some
columns/data fields would not be included in the analysis (this is from an
analysis contexts). Still for the purposes of other projects, field that I
may have said were not useful were those that are not clear. I suggest we
make certain fields clear to the user. For example the 'sets field' is
often blank or you find there are multiple 'sets' columns with different
number sequencing, 'Locale' column with 'en_us' data,'tags.0;, 'tags.1'
(both with different numbers). This examples are from an Uchaguzi CSV.
Personally, I do not understand what the data on these columns mean, maybe
some one else would. Perhaps, instead of removing fields as Shadrock says,
we should have some sort of key or codebook for someone to understand the
different fields.

**Comment by jrtricafort** [comment]
@willdoran FYI see @aoduor 's additional CSV feedback above.

**Comment by jrtricafort** [comment]
Moved to Cycle 7 - in Development, assigned to @willdoran

**Comment by Shadrock** [comment]
I think we're close to closing this out. @willdoran can you please create an export of posts, from both Uchaguzi deployments, by survey type? I realize that this will result in multiple .csv files per deployment and that only posts assigned to a survey will be displayed. But having them will help me make some comparisons to the large, single file that I have for each deployment and greatly speed up my evaluation process. You can place them in the same Drive folder as the others.

**Comment by willdoran** [comment]
@Shadrock Sure I can probably do that later today.

**Comment by willdoran** [comment]
@Shadrock those are in the drive

**Comment by Shadrock** [comment]
I've done a cursory examination of the data and they look ok (although I would really like to see these better structured for the average user) and am closing the issue. If there's anything exceedingly strange, I'll re-open it, but I think I can get most of what I need from a mix of Angie and grueling data clean up.

**Comment by Shadrock** [comment]
The data @willdoran provided were great... except that they did not include the Twitter ID for posts created directly (not manually by a volunteer). This causes two problems: the first is that it makes it very difficult to differentiate between a post purposefully sent to the deployment and a Tweet that was collected automatically and this differentiation is at the heart of the some of the impact we're trying to measure. Moreover, by archiving Tweet content and not providing Tweet IDs we effectively 'lock' the data set because COMRADES researchers cannot share it with others: see #3329 for more details on this. 

So what I am asking for in this issue is the same outputs that will created before, with the addition that the Twitter ID needs to be included for all Tweets. Also, _any information_ in the database that helps distinguish post sources (e.g. SMS, etc.) should be included.

**Comment by rowasc** [comment]
Thanks, @Shadrock. 
I need to deal with a small P0 thing then I'll go get this data for you, should be done today. I'll update you here .

**Comment by rowasc** [comment]
@Shadrock starting on this

**Comment by rowasc** [comment]
UPDATE, sent datasets, I think they are fine unless we want to remove more data than the unstructured text of tweets, ping me and we can adjust as needed

**Comment by rowasc** [comment]
Closing as this was delivered
